core/cas-server-core-events-api/src/main/java/org/apereo/cas/support/events/listener/CasAuthenticationAuthenticationEventListener.java

```diff
@@ -84,7 +82,7 @@ public void handleCasTicketGrantingTicketCreatedEvent(final CasTicketGrantingTic
 
     @Override
     public void handleCasTicketGrantingTicketDeletedEvent(final CasTicketGrantingTicketDestroyedEvent event) throws Exception {
-        val dto = prepareCasEvent(event);
+        val dto = prepareCasEvent(event,event.getClientInfo());
```

Comment:
If you are passing `event`, and `event.getClientInfo()` belongs to the same object, you don't need to pass the 2nd parameter.

Reply:
Until a decision is made on where to move the 'getter' into a super class of these events, this cannot be done.
I'm hesitant to add to AbstractCasEvent, but not opposed to create an intermediary.  The reason for the hesitancy is there is nothing preventing these events from being published from another 'Listener' of an event, which would cause the ClientInfo object to possibly have wrong data again.  I'm happy to hear your thoughts as you have more experience with where the AbstractCasEvent is being used.  If you don't see an issue with it existing there, I'm happy to add it there.

Reply:
I see what you mean. 

Your suspicion is well justified, and that is because the code is doing things like this:

```java
    public CasTicketGrantingTicketCreatedEvent(final Object source, final TicketGrantingTicket ticketGrantingTicket) {
        super(source, ticketGrantingTicket);
        clientInfo = ClientInfoHolder.getClientInfo();
    }
```

This should be more like this:

```java
    public CasTicketGrantingTicketCreatedEvent(final Object source, final TicketGrantingTicket ticketGrantingTicket, final ClientInfo clientInfo) {
        super(source, ticketGrantingTicket);
        this.clientInfo = clientInfo;
    }
```

The problem with the first approach is exactly as you described. A seemingly self-contained class relies on a semi-static data structure. If you let the caller provide the client-info, then it will/should be bound to the immediate thread context the caller executes, and not a static client-info thingy. 

This change would then allow you to move the field up to `AbstractCasEvent`. The repository accepts all `AbstractCasEvent` types, and so it would make sense for the field to stay up there. Of course, I realize the scope creep here, so please let me know if this is too much and I am happy to help you out in any way.

Reply:
I don't think it is.. I preferred to move it out as a parameter, but wasn't sure all the consequences. So I thought I would start small, and move out if need be.  I've patched what we are using, just didn't think we would want to re-implement the wheel every time we upgrade.  I think scope creep would be asking to fix other issues not related to this pull request, like using a timestamp for a unique ID rather than maybe a UUID for the cas-service-events-registry(at least in the mongodb instance), or storing creationTime as a string rather than a date in the mongodb instance of the repository which makes setting a TTL index rather difficult.  I'll get to this later this evening.  Thanks again for all the helpful feedback and discussion.

Reply:
Sounds good, thank you for sticking with this so far :)

Reply:
I have a question looking for some direction. I'm fairly close to being done with the logic change.  My question resolves around the CasConfiguration*Events and the CasRegisteredService*Events.  The CasConfigurationEvents are being created in a Watcher service, which I'm pretty sure isn't end-user initiated.  I believe some of the CasRegisteredService*Events are also being created in a similar type WatcherService.  At least one of these watcher services are being created at runtime during first usage, which is an acceptable thing to do.

The  calls originating from the watcher services probably aren’t ever going to be end-user created.  However, can you foresee issues with ClientInfoHolder.getClientInfo() not returning null in those cases?  

We have something similar in one of our code bases that injects username into service beans via the RequestContextHolder object and a static spring factory object, and I have seen it return null before, so I know its probable this would work as expected.  I’m just concerned if something in ‘user-land’ kicks off the bean creation, that the clientInfo object ends up being not null, but probably should be.  I could try and write a test to alleviate my concerns.. Just not sure the best way to approach that at the moment.  Am I just overthinking the whole situation?

If you think its fine, I’m happy to pass in the ClientInfo object from ClientInfoHolder.  If you think there may be issues, then I’m happy to create a 2nd constructor that can take ClientInfo, but leave the logic alone with the existing usages.

I’m aware that I’ll need to check the cas-management code-base for compile issues if I get replace the old constructors for the CasRegisteredService events.  Not sure if there is another Apereo owned project that could have a similar issue.  If there is, let me know, and I can check those out as well.

Anyway, happy to hear your thoughts.

Reply:
It's quite possible that you might get a ClientInfo that is null, yes. However I don't think this is a problem you need to worry about. This (the capturing and recording of events) happens on a best-effort basis and there are null checks in place to make sure we only record things we can record, and ignore the stuff we cannot. I'd proceed.